# Data Cleaning & Exploratory Analysis - Global Superstore Datasets

## Importar bibliotecas:

In [1]:
from IPython.display import display, Markdown

## Leer el archivo:

In [2]:
import pandas as pd

# Cargamos el dataset original 
try:
    ruta = 'https://raw.githubusercontent.com/3rick-CR/Global-Sales-Analytics-System/refs/heads/main/data/raw/Global_Superstore.csv'
    df = pd.read_csv(ruta, encoding='latin1')
    
    print("Se cargo el archivo")
    
except FileNotFoundError:
    print("Error: No encontré el archivo CSV. Revisa que esté dentro de la carpeta 'data'.")
except Exception as e:
    print(f"Ocurrió un error inesperado: {e}")

Se cargo el archivo


## Visualizar filas y columnas:

In [3]:
print(f"Filas detectadas: {df.shape[0]}")
print(f"Columnas detectadas: {df.shape[1]}")
display(df.head())

Filas detectadas: 51290
Columnas detectadas: 27


,Category,City,Country,Customer ID,Customer Name,Discount,Market,记录数,Order Date,Order ID,...,Sales,Segment,Ship Date,Ship Mode,Shipping Cost,State,Sub-Category,Year,Market2,weeknum
0,Office Supplies,Los Angeles,United States,LS-172304,Lycoris Saunders,0.0,US,1,2011-01-07 00:00:00.000,CA-2011-130813,...,19,Consumer,2011-01-09 00:00:00.000,Second Class,4.37,California,Paper,2011,North America,2
1,Office Supplies,Los Angeles,United States,MV-174854,Mark Van Huff,0.0,US,1,2011-01-21 00:00:00.000,CA-2011-148614,...,19,Consumer,2011-01-26 00:00:00.000,Standard Class,0.94,California,Paper,2011,North America,4
2,Office Supplies,Los Angeles,United States,CS-121304,Chad Sievert,0.0,US,1,2011-08-05 00:00:00.000,CA-2011-118962,...,21,Consumer,2011-08-09 00:00:00.000,Standard Class,1.81,California,Paper,2011,North America,32
3,Office Supplies,Los Angeles,United States,CS-121304,Chad Sievert,0.0,US,1,2011-08-05 00:00:00.000,CA-2011-118962,...,111,Consumer,2011-08-09 00:00:00.000,Standard Class,4.59,California,Paper,2011,North America,32
4,Office Supplies,Los Angeles,United States,AP-109154,Arthur Prichep,0.0,US,1,2011-09-29 00:00:00.000,CA-2011-146969,...,6,Consumer,2011-10-03 00:00:00.000,Standard Class,1.32,California,Paper,2011,North America,40


## Normalizar nombres de columna:

In [4]:
#Quita espacios , pasa el texto a minuscula y remplaza valores 
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('-', '_')

print("Nombres de columnas:")
print(df.columns.tolist())

Nombres de columnas:
['category', 'city', 'country', 'customer_id', 'customer_name', 'discount', 'market', 'è®°å½\x95æ\x95°', 'order_date', 'order_id', 'order_priority', 'product_id', 'product_name', 'profit', 'quantity', 'region', 'row_id', 'sales', 'segment', 'ship_date', 'ship_mode', 'shipping_cost', 'state', 'sub_category', 'year', 'market2', 'weeknum']


## Buscar valores nulos:

In [5]:
# Contamos cuántos nulos hay por columna
nulos = df.isnull().sum()

#visualizamos si existen valores NaN
print("--- Columnas con valores (NaN) ---")
for columna, cantidad in nulos.items():
    if cantidad > 0:
        print(f"La columna '{columna}' tiene {cantidad} valores nulos.")





--- Columnas con valores (NaN) ---


* **Nota:** En esta ocasion el dataset de kaggle viene sin valores nulos, como ejemplo en caso de haberlos tenido podriamos utilizar el metodo df = df.dropna(subset=['Sales', 'Profit']) para eliminar unicamente las filas de las columnas 'Sales', 'Profit' que contengan valores nulos.

## Eliminacion de columnas:

In [6]:
#(columnas que no aportan al analisis)

# Valores unicos de las columna de interes
print(f"{'记录数'}:{df['记录数'].unique()}")

记录数:[1]


In [7]:
#borrado
df.drop(columns=['记录数'], inplace=True)
df.drop(columns=['row_id'], inplace = True)
df.drop(columns=['weeknum'], inplace = True)

#comprobacion
print(df.columns)

Index(['category', 'city', 'country', 'customer_id', 'customer_name',
       'discount', 'market', 'order_date', 'order_id', 'order_priority',
       'product_id', 'product_name', 'profit', 'quantity', 'region', 'sales',
       'segment', 'ship_date', 'ship_mode', 'shipping_cost', 'state',
       'sub_category', 'year', 'market2'],
      dtype='object')


### Observaciones de la Limpieza de Datos:

Tras el análisis inicial de las **51,290 filas**, detecté lo siguiente:
* **Valores Nulos:** El dataset no contenia valores nulos.

* **Formato de Columnas:** Estandaricé los nombres a *snake_case* (minúsculas y guiones bajos) para evitar errores de sintaxis en el futuro.

* **Columnas que se eliminaron:** 

     记录数: Esta columna de nombre atipico solo almacenaba un unico valor siendo este el valor de 1, se decidio eliminar la columna ya que no era significativa para el análisis

    row_id: Se elimino esta columna debido a que el dataframe cuenta con varios identificadores que consideramos son mas importantes como order_id

    weeknum: El dataframe cuenta con columna de las fechas que corresponden al dia que se realizo la compra, por este motivo nos resulto redundante tener esta columna de nombre weeknum

    
* **Tipos de Datos:** Se convirtieron las fechas de "texto" a "datetime", lo que nos permitirá analizar la estacionalidad de las ventas más adelante.

## Convertir los datos de columna al tipo de dato correcto:

In [8]:
#Ver el tipo de dato por columna
print(df.dtypes.to_string())

category           object
city               object
country            object
customer_id        object
customer_name      object
discount          float64
market             object
order_date         object
order_id           object
order_priority     object
product_id         object
product_name       object
profit            float64
quantity            int64
region             object
sales               int64
segment            object
ship_date          object
ship_mode          object
shipping_cost     float64
state              object
sub_category       object
year                int64
market2            object


In [9]:
#Covertir al tipo de dato correcto 

# Fechas de texto" (object) a "fecha" (datetime) 
df['order_date'] = pd.to_datetime(df['order_date'])
df['ship_date'] = pd.to_datetime(df['ship_date'])
df['sales'] = df['sales'].astype(float)

## Creación de nuevos Indicadores Clave KPIs: 

In [10]:

#Genera una columna que calcula el tiempo que tardo en enviarse el producto
df['days_to_ship'] = (df['ship_date'] - df['order_date']).dt.days

display(Markdown("### Tiempos de Envío (Logística):"))
display(df[['order_date', 'ship_date', 'days_to_ship']].head())

### Tiempos de Envío (Logística):

,order_date,ship_date,days_to_ship
0,2011-01-07,2011-01-09,2
1,2011-01-21,2011-01-26,5
2,2011-08-05,2011-08-09,4
3,2011-08-05,2011-08-09,4
4,2011-09-29,2011-10-03,4


In [11]:
#Columna que muestra el valor de los descuentos
df['discount_amount'] = df['sales'] * df['discount']

display(Markdown("### Cálculo de Descuento en Moneda Real:"))
display(df[['sales', 'discount', 'discount_amount']].head())

### Cálculo de Descuento en Moneda Real:

,sales,discount,discount_amount
0,19.0,0.0,0.0
1,19.0,0.0,0.0
2,21.0,0.0,0.0
3,111.0,0.0,0.0
4,6.0,0.0,0.0


In [12]:
#Columna que muestra cual fue el precio unitario 
df['unit_price'] = df['sales'] / df['quantity']

display(Markdown("### Precio unitario:"))
display(df[['unit_price', 'sales', 'quantity']].head())

### Precio unitario:

,unit_price,sales,quantity
0,6.333333,19.0,3
1,9.500000,19.0,2
2,7.000000,21.0,3
3,55.500000,111.0,2
4,6.000000,6.0,1


In [13]:
#Columna de categorizacion o etiquetas segun el desempeño
df['profit_status'] = df['profit'].apply(lambda x: 'Profit' if x > 0 else ('Loss' if x < 0 else 'Break-even'))

display(Markdown("### Validación de Categorización de Desempeño:"))
display(df[['sales', 'profit', 'profit_status']].head(10))

### Validación de Categorización de Desempeño:

,sales,profit,profit_status
0,19.0,9.3312,Profit
1,19.0,9.2928,Profit
2,21.0,9.8418,Profit
3,111.0,53.2608,Profit
4,6.0,3.1104,Profit
5,13.0,6.5856,Profit
6,19.0,9.3312,Profit
7,12.0,5.8604,Profit
8,54.0,24.2190,Profit
9,49.0,23.0864,Profit


In [14]:
display(Markdown("#### Auditoría Final del Dataset (Pre-Exportación)"))

print("Estructura técnica de las columnas:")
display(Markdown("---"))

count = 0
for key in df.keys():
    count += 1
    print(f"{count}) {key}: {df[key].dtype}")



# Un vistazo a las últimas columnas creadas
#display(df.tail())

#### Auditoría Final del Dataset (Pre-Exportación)

Estructura técnica de las columnas:


---

1) category: object
2) city: object
3) country: object
4) customer_id: object
5) customer_name: object
6) discount: float64
7) market: object
8) order_date: datetime64[ns]
9) order_id: object
10) order_priority: object
11) product_id: object
12) product_name: object
13) profit: float64
14) quantity: int64
15) region: object
16) sales: float64
17) segment: object
18) ship_date: datetime64[ns]
19) ship_mode: object
20) shipping_cost: float64
21) state: object
22) sub_category: object
23) year: int64
24) market2: object
25) days_to_ship: int64
26) discount_amount: float64
27) unit_price: float64
28) profit_status: object


In [15]:
display(Markdown("Dimensiones del dataset:"))

filas = df.shape[0]
columnas = df.shape[1]

print(f"{filas} filas y {columnas} columnas")

Dimensiones del dataset:

51290 filas y 28 columnas


## Exportar el csv limpio:

In [16]:

df.to_csv('Global_Superstore_Clean.csv', index=False, encoding='utf-8')
print("Se genero el csv existosamente.")

Se genero el csv existosamente.
